src/mjolnir/adminbuilder.cc

```diff
@@ -371,114 +508,91 @@ void BuildAdminFromPBF(const boost::property_tree::ptree& pt,
-          } else {
-            sqlite3_bind_null(stmt, 5);
-          }
-
-          sqlite3_bind_int(stmt, 6, admin.drive_on_right());
-          sqlite3_bind_int(stmt, 7, admin.allow_intersection_names());
-          sqlite3_bind_text(stmt, 8, wkt.c_str(), wkt.length(), SQLITE_STATIC);
-          /* performing INSERT INTO */
-          ret = sqlite3_step(stmt);
-          if (ret == SQLITE_DONE || ret == SQLITE_ROW) {
-            continue;
-          }
-          LOG_ERROR("sqlite3_step() error: " + std::string(sqlite3_errmsg(db_handle)));
-          LOG_ERROR("sqlite3_step() Name: " +
-                    osm_admin_data.name_offset_map.name(admin.name_index()));
-          LOG_ERROR("sqlite3_step() Name:en: " +
-                    osm_admin_data.name_offset_map.name(admin.name_en_index()));
-          LOG_ERROR("sqlite3_step() Admin Level: " + std::to_string(admin.admin_level()));
-          LOG_ERROR("sqlite3_step() Drive on Right: " + std::to_string(admin.drive_on_right()));
-          LOG_ERROR("sqlite3_step() Allow Intersection Names: " +
-                    std::to_string(admin.allow_intersection_names()));
-        }
-      } // has data
-    }   // admins
-  } catch (std::exception& e) {
-    LOG_ERROR("Standard exception processing relation: " + std::string(e.what()));
-  } catch (...) { LOG_ERROR("Exception caught processing relations."); }
+  for (const auto& admin : admin_data.admins) {
+    std::pair<std::string, uint64_t> admin_info(admin_data.name_offset_map.name(admin.name_index),
+                                                admin.id);
+    LOG_DEBUG("Building admin: " + admin_info.first);
+
+    // do inners and outers separately
+    bool complete = true;
+    std::array<std::vector<ring_t>, 2> outers_inners;
+    for (bool outer : {true, false}) {
+      // grab the ring segments and a lookup to find them when connecting them
+      std::vector<ring_t> lines;
+      std::unordered_multimap<valhalla::midgard::PointLL, size_t> line_lookup;
+      if (!to_segments(admin_data, admin, admin_info.first, outer, lines, line_lookup)) {
+        complete = false;
+        break;
+      }
+      // connect them into a series of one or more rings
+      to_rings(admin_info, lines, line_lookup, outers_inners[!outer], outers_inners[1]);
+    }
+
+    // if we didn't have a complete relation (ie some members were missing) we bail
+    if (!complete || outers_inners.front().empty()) {
+      LOG_WARN(admin_info.first + " (" + std::to_string(admin_info.second) +
+               ") is degenerate and will be skipped");
+      continue;
+    }
+
+    // convert the rings into multipolygons
+    auto multipolygon = to_multipolygon(admin_info, outers_inners.front(), outers_inners.back());
+
+    // convert that into wkt format so we can put it into sqlite
+    std::stringstream ss;
+    ss << std::setprecision(7) << boost::geometry::wkt(multipolygon);
```

Comment:
this was the last bug. nowhere in the boost docs does it say about `setprecsion` for the  `wkt` stream. the result is a pile of truncated coords with geometries that look like crap. similar report here: https://stackoverflow.com/questions/28083952/boost-geometry-precision-loss